# PAQ(Personally Asked Questions)

## 목차
1. 매개변수 self, cls 차이, 그리고 static method
2. 더블 언더스코어: Magic Method
3. mutable과 hashable에 대한 이해
4. classmethod와 staticmethod의 차이
5. Observer 패턴에 대한 이해
6. Super() 기본 개념
7. queue 클래스
8. thread에 대한 이해
9. 다시 봐도 헷갈리는 비트연산자
10. 논리 연산자(and, or, not)

# 1. 매개변수 self, cls 차이, 그리고 static method
- 출처: https://paphopu.tistory.com/entry/Python-%EB%A7%A4%EA%B0%9C%EB%B3%80%EC%88%98-self-%EC%99%80-cls%EC%9D%98-%EC%B0%A8%EC%9D%B4-%EA%B7%B8%EB%A6%AC%EA%B3%A0-static-method%EC%97%90-%EB%8C%80%ED%95%B4%EC%84%9C

## 1.1 Instance Method
- Instance Method는 클래스 내부에 정의되어 있는 함수 호출시, Instance(객체)를 필요로 한다는 조건 존재
- 이때 첫번째 매개변수는 항상 self

In [5]:
# 인스턴스 메소드 예제
class InstMethod:
    
    def __init__(self):
        self.name = 'Seyoung Nam'
        
    def print_name(self):
        print('My name is %s' %(self.name))
        
# 인스턴스 선언
name_instance = InstMethod()

# print_name이라는 함수를 호출하기 위해 name_instance 같은 객체 정의 필요
name_instance.print_name()

My name is Seyoung Nam


## 1.2 Class Method
- Class Method는 인스턴스 선언이 필요없이 "클래스.함수명"으로 바로 불러올 수 있음
- 자기 자신을 첫번째 매개변수로 받기에 self를 사용하지 않고 cls를 사용

In [6]:
# 클래스 메소드 예제
class ClassMethod:
    
    @classmethod
    def print_name(cls):
        print('my name is %s' %(cls.__class__.__name__))
        
# 객체를 따로 선언해 줄 필요없이 함수 호출 가능
ClassMethod.print_name()

my name is type


In [8]:
# 클래스 메소드라도 인스턴스 선언해서 호출 또한 가능 
class ClassMethod2:
    name = 'Seyoung'
    
    @classmethod
    def print_name(cls):
        print('my nam is %s.' %(cls.name))
        
ins = ClassMethod2()
ins.print_name()

my nam is Seyoung.


## 1.3 Static Method
- Static Method는 앞의 두 method들과 다르게 인스턴스나 클래스를 인자로 받지 않음
- Static Method는 클래스 내부에 선언되어 클래스 네임스페이스 안에 저장되는 점
- 그외에는 일반 method들과 크게 다른 점 없음
- Static method의 특징은 클래스를 통해서도, 인스턴스를 통해서도 호출이 가능하다는 점

In [9]:
class StaticMethod:
    
    @staticmethod
    def print_name(name):
        return '내 이름은 {} 입니다.'.format(name)
    
# 클래스를 통해서 호출가능
print(StaticMethod.print_name('Seyoung'))

# 인스턴스를 통해서도 호출 가능
me = StaticMethod()
print(me.print_name('NSY'))

내 이름은 Seyoung 입니다.
내 이름은 NSY 입니다.


# 2. 더블 언더스코어: Magic Method
- 출처 : https://corikachu.github.io/articles/python/python-magic-method
- 파이썬에서는 미리 정의되어 있는 특별한 이름을 가진 메소드들을 재정의 가능
- 미리 정해져 있는 메소드, 즉 Built-in 함수들을 Special Method 혹은 Double Underscore Method라 함
- 공식문서: https://docs.python.org/3/reference/datamodel.html#special-method-names

## 2.1 객체의 생성과 초기화
- \_\_new\_\_(cls[, ...]): 새 인스턴스 만들 때 클래스를 불러오기 위해 제일 처음으로 실행되는 메소드. super().\_\_new\_\_(cls[, ...]) 사용 
- \_\_init\_\_(self[, ...]): \_\_new\_\_ 이후, caller에게 return 전에 호출. 불러온 클래스를 바탕으로 해당 클래스의 신규 인스턴스를 만듬. self는 instance를 의미.
- In sum, \_\_new\_\_() to create it, and \_\_init\_\_() to customize it
- \_\_del\_\_(self): 객체가 소멸될 때 호출, 객체 소멸시 해야할 일을 지정

In [10]:
class NumBox:
    
    def __new__(cls, *args, **kwargs):
        if len(args) < 1: # 인자가 들어오지 않은 경우
            return None
        else:
            return super(NumBox, cls).__new__(cls)  # object를 반환
        
    def __init__(self, num=None):
        self.num = num  # 받은 인자 num을 인스턴스 변수로 지정
        
    def __repr__(self):
        return str(self.num)

In [11]:
a = NumBox()  # 인자 없이 객체 생성
type(a)

NoneType

In [12]:
b = NumBox(10)  # 인자 넣고 객체 생성
b # __repr__()에 의해 self.num 프린트

10

In [13]:
type(b)

__main__.NumBox

## 2.2 객체의 표현
- \_\_repr\_\_(self): 객체를 나타내는 __공식적인__ 문자열. built-in 함수인 repr()로 호출, 반환값 string, \_\_str\_\_()과 달리 좀 더 명확함을 지향.
- \_\_str\_\_(self): 객체를 나타내는 비공식적인 문자열, 객체를 이해하기 쉽게 표현. \__repr__보다 보기 쉬운 문자열을 출력하는 것에 지향점, string 문자열 반환. exception 없이 a more convenient or concise representation can be used.라는 문구 뜸
- \_\_bytes\_\_(self): 객체의 byte 문자열을 리턴, bytes()로 호출
- \_\_format\_\_(self): 객체의 "formatted" string representation 생성, format()으로 호출

In [14]:
class StrBox(object):
    
    def __init__(self, string):
        self.string = string
        
    def __repr__(self):
        return "A('{}')".format(self.string)
    
    def __bytes__(self):
        return str.encode(self.string)
    
    def __format__(self, format):
        if format == 'this-string':
            return "This string: {}".format(self.string)
        return self.string

In [15]:
a = StrBox('Life is short, you need python')
a  #  __repr__()

A('Life is short, you need python')

In [16]:
repr(a)

"A('Life is short, you need python')"

In [17]:
str(a)  # str 정의되어 있지 않으면 repr과 동일

"A('Life is short, you need python')"

In [18]:
bytes(a)

b'Life is short, you need python'

In [21]:
"{:this-string}".format(a)

'This string: Life is short, you need python'

## 2.3 속성 관리
- \_\_getattr\_\_(self, name): 객체의 __없는__ 속성을 참조하려 할때 호출. 찾는 속성이 있다면 호출X
- \_\_getattribute\_\_(self, name): 객체 속성 호출시 무조건 호출. 이 메소드가 재정의 되어 있다면 \_\_getattr\_\_는 호출되지 않으므로 명시적으로 호출해야하거나 AttributeError 에러를 발생시켜야 함
- \_\_setattr\_\_(self, name, value): 객체의 속성을 변경할때 호출. 
- \_\_delattr\_\_(self, name): 객체의 속성을 del 키워드로 지울 때 호출
- \_\_dir\_\_(self): 객체가 가지고 있는 모든 속성들을 보여주는 dir()을 사용할때 호출
- \_\_slots\_\_: 사용할 변수의 이름을 미리 지정할 수 있음.

In [23]:
# __getattr__(self, name) 예제

class LazyDB:
    
    def __init__(self):
        self.exists = 5
    
    def __getattr__(self, name):
        value = 'Value for %s' % name
        setattr(self, name, value)
        return value
    
data = LazyDB()

print('Before: ', data.__dict__)
print('foo ', data.foo)
print('After: ', data.__dict__)

Before:  {'exists': 5}
foo  Value for foo
After:  {'exists': 5, 'foo': 'Value for foo'}


In [26]:
class NameBox:

    person_name = "Seyoung"
    def __getattr__(self, name): # 없는 속성일 때 자동 호출
        print("Not Found: {}".format(name))

    def __setattr__(self, name, value):        
        print("Set attribute: {} is {}".format(name, value))
        super().__setattr__(name, value)
#         self.person_name = value  # 재귀적으로 호출되어 문제가 있는 부분

In [27]:
box = NameBox()
box.person_name

'Seyoung'

In [28]:
box.name

Not Found: name


In [29]:
box.name = "Say yes"

Set attribute: name is Say yes


In [30]:
dir(box)  # box.name이 있는지 확인

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'name',
 'person_name']

## 2.4 Descriptors 관리
### Descriptor는 \_\_get\_\_(), \_\_set\_\_(), \_\_delete\_\_() 메소드로 구성된 프로토콜을 구현한 클래스
- \_\_get\_\_(self, instance, owner): 특정 오브젝트의 값을 참조할때 호출
- \_\_set\_\_(self, instance, value): 특정 오브젝트의 값을 변경할때 호출
- \_\_delete\_\_(self, instance): 특정 오브젝트의 값을 삭제할때 호출

In [31]:
class Rating():
    def __init__(self, rating=3):
        self.rating = rating
        
    def __set__(self, instance, value):
        if value < 0 or value > 5:
            raise ValueError('rating must be 0~5')
        else:
            setattr(instance, 'rating', value)
            
    def __get__(self, instance, owner):
        return getattr(instance, 'rating')
    
class MovieReview():
    story = Rating()
    acting = Rating()
    fun = Rating()

In [32]:
a = MovieReview()
a.story = 10

ValueError: rating must be 0~5

In [33]:
a.acting = 2

In [34]:
a.acting

2

## 2.5 컨테이너 관리: 콜렉션과 반복
### 컨테이너는 list와 tuple 같은 sequence와 dict같은 mapping을 뜻함
- \_\_len\_\_(self): 객체의 길이 반환, len()으로 호출
- \_\_length_hint\_\_(self): 객체의 대략적인 측정 길이 반환, operator.length_hint()로 호출
- \_\_getitem\_\_(self, key): 객체에서 [] 연산자를 사용하여 조회시 동작을 정의. 즉, \_list[10]은 \_list.\_\_getitem\_\_(10)으로 동작. 키 타입이 적절치 않으면 TypeError, 키가 인덱스 벗어나면 IndexError를 던짐
- \_\_missing\_\_(self, key): 키가 dict에 없을 경우 호출
- \_\_setitem\_\_(self, key, value): 객체에서 [] 연산자를 사용해서 변수 지정할 때 동작을 정의. 즉, \_list[10] = 1은 \_list.\_\_setitem\_\_(10, 1)으로 동작
- \_\_delitem\_\_(self, key): del object[]를 사용하는 경우 동작을 정의
- \_\_iter\_\_(self): 컨테이너의 iterator를 반환
- \_\_reversed\_\_(self): 순서가 반대로 바뀌는 함수인 reversed()로 호출
- \_\_contains\_\_(self, item): item이 존재한다면 True, 그렇지 않으면 False를 반환하는 메소드 정의

## 2.6 나머지 클래스 서비스들
- \_\_prepare\_\_(metacls, name, bases, \*\*kwds): 메타 클래스 네임스페이스에 대한 dictionary를 만듬. 메타 클래스에 이 속성이 없다면 빈 dict()로 초기화
- \_\_instancecheck\_\_(self, instance): 클래스의 인스턴스이면 참을 반환, isinstance(instance, class)로 호출
- \_\_subclasscheck\_\_(self, subclass): 클래스의 서브클래스라면 참을 반환, issubclass(subclass, class)로 호출

## 2.7 연산
### 2.7.1 단항 연산자
- \_\_neg\_\_(self): -object를 정의
- \_\_pos\_\_(self): +object를 정의
- \_\_abs\_\_(self): abs()를 정의
- \_\_invert\_\_(self): 비트 연산 ~object를 정의

### 2.7.2 비교 연산자
- object.\_\_lt\_\_(self, other) : x<y를 정의
- object.\_\_le\_\_(self, other) : x<=y를 정의
- object.\_\_eq\_\_(self, other) : x==y를 정의
- object.\_\_ne\_\_(self, other) : x!=y를 정의
- object.\_\_gt\_\_(self, other) : x>y를 정의
- object.\_\_ge\_\_(self, other) : x>=y를 정의

In [2]:
# 예제 문자열 길이 비교도 가능
class StrBox(str):
    def __new__(cls, string):
        return str.__new__(cls, string)
    def __lt__(self, other):
        return len(self) < len(other)
    def __le__(self, other):
        return len(self) <= len(other)
    def __gt__(self, other):
        return len(self) > len(other)
    def __ge__(self, other):
        return len(self) >= len(other)
    def __eq__(self, other):
        return len(self) == len(other)
    def __ne__(self, other):
        return len(self) != len(other)

In [4]:
a = StrBox('seyoung')
b = StrBox('saekwang')
a>b

False

### 2.7.3 산술 연산자
- 예시: \_\_add\_\_는 + 연산에 대해 정의

In [5]:
class NumBox:
    
    def __init__(self, num):
        self.number = num
        
    def __add__(self, num):
        return NumBox(self.number + num)

In [13]:
a = NumBox(10)
a+10

In [14]:
10+a # 역순 연산자는 __radd__를 통해 정의, 복합연산자(+=)는 __iadd__를 통해 정의

TypeError: unsupported operand type(s) for +: 'int' and 'NumBox'

- \_\_add\_\_(self, other): x+y 연산을 정의, \_\_radd\_\_는 역순 연산자, \_\_iadd\_\_는 복합 할당 연산자
- \_\_sub\_\_(self, other): x-y 연산을 정의, \_\_rsub\_\_는 역순 연산자, \_\_isub\_\_는 복합 할당 연산자
- \_\_mul\_\_(self, other): x\*y 연산을 정의, \_\_rmul\_\_는 역순 연산자, \_\_imul\_\_는 복합 할당 연산자
- \_\_matmul\_\_(self, other): x@y 연산을 정의, \_\_rmatmul\_\_는 역순 연산자, \_\_imatmul\_\_는 복합 할당 연산자
- \_\_truediv\_\_(self, other): x/y 연산을 정의, \_\_rtruediv\_\_는 역순 연산자, \_\_itruediv\_\_는 복합 할당 연산자
- \_\_floordiv\_\_(self, other): x//y 연산을 정의, \_\_rfloordiv\_\_는 역순 연산자, \_\_ifloordiv\_\_는 복합 할당 연산자
- \_\_mod\_\_(self, other): x%y 연산을 정의, \_\_rmod\_\_는 역순 연산자, \_\_imod\_\_는 복합 할당 연산자
- \_\_divmod\_\_(self, other): divmod()를 통해 호출되는 연산을 정의
- \_\_pow\_\_(self, other[, modulo]): x\*\*y 연산을 정의, pow()을 통해서도 호출
- \_\_round\_\_(self[, n]): 반올림 함수 round()를 통해 호출되는 연산을 정의

### 2.7.4 비트 연산자와 논리 연산자
- \_\_lshift\_\_(self, other): x << y 시프트 연산을 정의. \_\_rlshift\_\_는 역순 연산자, \_\_ilshift\_\_는 복합 할당 연산자
- \_\_rshift\_\_(self, other): x >> y 시프트 연산을 정의. \_\_rrshift\_\_는 역순 연산자, \_\_rlshift\_\_는 복합 할당 연산자
- \_\_and\_\_(self, other): x & y 연산을 정의. \_\_rand\_\_는 역순 연산자, \_\_iand\_\_는 복합 할당 연산자
- \_\_or\_\_(self, other): x | y 연산을 정의. \_\_ror\_\_는 역순 연산자, \_\_ior\_\_는 복합 할당 연산자
- \_\_xor\_\_(self, other): x^y 연산을 정의. \_\_rxor\_\_는 역순 연산자, \_\_ixor\_\_는 복합 할당 연산자

## 2.8 타입 변환
- \_\_int\_\_(self): 정수 변환 함수 int()를 통해 호출되는 연산을 정의
- \_\_float\_\_(self): 실수 변환 함수 float()를 통해 호출되는 연산을 정의
- \_\_complex\_\_(self): 복소수 변환 함수 complex()를 통해 호출되는 연산을 정의
- \_\_bool\_\_(self): bool()을 통해 호출되는 연산을 정의. True나 False를 반환.
- \_\_hash\_\_(self): hash()를 통해 호출되는 연산을 정의. 정수를 반환.
- \_\_index\_\_(self): slice expression에 객체가 사용될때 사용할 정수 형태를 정의.

In [15]:
# __index__ 관련 예제
class Slice:
    def __index__(self):
        return 1

In [16]:
slice = Slice()
_list = ["123", "456", "789"]
_list[slice]

'456'

## 2.9 매직 메소드 다루기 - 컨텍스트 매니저
with 키워드를 통해 블럭에 진입할 때, 컨텍스트 매니저를 통해서 시작과 끝에 할 일을 처리
- \_\_enter\_\_(self): with로 블럭에 진입할 때 해야할 일을 정함
- \_\_exit\_\_(self, exc_type, exc_value, traceback): 블랙이 끝날때 해야할 일을 정함. exception이 발생한 경우에도 호출. 정상적으로 종료되었다면 exc_type, exc_value, traceback은 None로 들어옴.

In [17]:
# socket을 자동으로 닫아주는 SocketWrapper. 
# with SocketWrapper() as so와 같은 코드로 사용가능하며,
# with 블럭이 끝나면 자동으로 __exit__를 호출해서 소켓 연결을 닫음
import socket

class SocketWrapper:
    
    def __init__(self):
        self.so = socket.socket(socket.AF_INET, socket.SOCK_STREAM)  # 새로운 소켓을 생성
        
    def __enter__(self):
        self.so.connect(('localhost', 8888))  # 소켓을 connect 하고
        return self.so  # 반환한다
    
    def __exit__(self, exception_type, exception_val, trace):
        try:
            self.so.close()  # with 구문이 끝나면 소켓을 닫는다
        except socket.error as msg:
            print(msg)

## 2.10 매직 메소드 다루기 - 비동기
- \_\_await\_\_(self): await 표현을 사용할 수 있는(awaitable) 객체를 만드는데 사용. iterator를 반환.
- \_\_aiter\_\_(self): 비동기를 위한 \_\_iter\_\_ 임. 비동기 iterator 반환.
- \_\_anext\_\_(self): 비동기를 위한 \_\_next\_\_임. awaitable한 결과를 반환. iteration이 끝나면 StopAsyncIteration 에러를 던짐
- \_\_aenter\_\_(self): 비동기 컨텍스트 매니저를 위한 메소드. \_\_enter\_\_랑 같음. awaitable 객체 반환.
- \_\_aexit\_\_(self, exc_type, exc_value, traceback): 비동기 컨텍스트 매니저를 위한 메소드. \_\_exit\_\_와 같음. awaitable 객체를 반환.

In [18]:
import asyncio

class AsyncIterator:
    def __init__(self, obj):
        self.obj = obj
        
    async def __aiter__(self):
        return self
    
    async def __anext__(self):
        try:
            return next(self.obj)
        except StopIteration:  # iteration이 끝일 경우
            raise StopAsyncIteration
            
async def example():
    _map = map(lambda x: x*2, [1,2,3])
    _iter = AsyncIterator(_map)
    async for x in _iter:
        print(x, end= ' ')
        
loop = asyncio.get_event_loop()
loop.run_until_complete(example())

RuntimeError: This event loop is already running

# 3. mutable과 hashable에 대한 이해

http://net-informations.com/python/iq/unhashable.htm

## 3.1 Hashing 이란?
- 해싱은 컴퓨터공학 컨셉 중 하나로 high performance, pseudo random access structures를 만드는 데 사용되며, 대량의 데이터가 빠르게 저장 및 전송을 가능케 함
- 해시(hash)는 어떤 종류의 값을 받아서 정수를 반환하는 함수를 말한다. 예를 들어 "Java"라는 string을 쓰면 key에 해당하는 어떤 integer를 만들고 이 key에 "Java"라는 value를 저장한다.
- key와 value는 항상 1대1 매칭되어야 하기 때문에 key는 value에 종속적이며, 따라서 value가 바뀌지 않는 이상 key도 바뀌면 안된다.
- 따라서 key의 객체 타입은 절대불변의 한가지 값만 가지는 불변객체(Immutable objects)

## 3.2 Hashable vs Unhashable
- 불변객체(Immutable objects)만이 hashable type으로 인정
- hashable objects: 숫자형(int, float, decimal, complex), string, tuple, range, frozenset, bytes
- unhasable objects: list, dict, set, bytearray, user-defined classes

결론: immutable과 hashable 객체가 짝!

In [1]:
# mutable한 객체를 unhashable 객체에 사용할 때 에러 발생
my_dict = {'name': 'john', [1,2,3]:'values'} # dict 내 key는 immutable 객체를 써야 하는데 mutable인 list를 넣음
my_dict # 에러 발생

TypeError: unhashable type: 'list'

In [2]:
my_dict = {'name': 'John', tuple([1,2,3]):'values'}
my_dict

{'name': 'John', (1, 2, 3): 'values'}

# 4. classmethod와 staticmethod의 차이

https://medium.com/@hckcksrl/python-%EC%A0%95%EC%A0%81%EB%A9%94%EC%86%8C%EB%93%9C-staticmethod-%EC%99%80-classmethod-6721b0977372

https://hamait.tistory.com/635


## 핵심내용
- 정적메소드는 클래스에서 직접 접근할 수 있는 메소드이다.
- 파이썬에서 클래스에서 직접 접근할 수 있는 메소드가 @staticmethod 와 @classmethod 두가지가 있다.
- 파이썬에서는 정적메소드임에도 불구하고 인스턴스에서도 접근이 가능하다.
- classmethod는 상속받은 클래스의 데이터 요소를 반환하는 반면, staticmethod는 상속 전 원 클래스의 데이터 요소를 반환하는 차이가 존재.

## 4.1 classmethod
- classmethod는 @classmethod라는 데코레이터 정의를 사용하여 정의
- instance method와 달리 self라는 인자 대신 cls라는 인자를 가짐

In [2]:
class Calc:
    
    @classmethod
    def add(cls, a, b):
        return a + b
    
cal = Calc()
cal.add(1,2)  # return 3

3

## 4.2 staticmethod
- @staticmethod라는 데코레이터를 사용하여 정의
- 인스턴스 메소드와 달리 self라는 인자가 없음

In [3]:
class Calc:
    
    @staticmethod
    def add(a ,b):
        return a + b

cal = Calc()
cal.add(1,2)  # return 3

3

## 4.3 @classmethod와 @staticmethod의 차이
두 정적 메소드는 상속에서 차이가 난다

In [8]:
class Person:
    default = '아빠'
    
    def __init__(self):
        self.data = self.default
        
    @classmethod
    def class_person(cls):
        return cls()
    
    @staticmethod
    def static_person():
        return Person()
    
class WhatPerson(Person):
    default = "엄마"
    
person1 = WhatPerson.class_person()   # return 엄마
person2 = WhatPerson.static_person()  # return 아빠
print(person1.data, person2.data)

엄마 아빠


staticmethod의 경우 부모 클래스의 클래스 속성 값을 가져오지만 classmethod의 경우 cls인자를 활용하여 현재 클래스의 클래스 속성을 가져옴

In [9]:
class Person:
    default = '아빠'
    
    def __init__(self):
        self.data = self.default

    
class WhatPerson(Person):
    default = "엄마"
    
    @classmethod
    def class_person(cls):
        return cls()
    
    @staticmethod
    def static_person():
        return Person()
    
person1 = WhatPerson.class_person()   # return 엄마
person2 = WhatPerson.static_person()  # return 아빠
print(person1.data, person2.data)

엄마 아빠


# 5. Observer 패턴에 대한 이해

https://www.youtube.com/watch?v=87MNuBgeg34

- 여러 사람에게 공지를 한번에 띄우거나 변화를 한번에 알릴 때 주로 쓰임(pattern: one speaking to many)
- 여기서 여러 사람(many to be notified)을 Observer, 공지를 띄우는 객체(one)를 Observable이라고 일컬음
- Observer를 Subscriber로, Observable를 Publisher로 보면 더 이해가 쉬움


In [1]:
# Simple Observer 

class Subscriber:   # 모든 subscriber는 메세지를 받는 update 함수를 가지고 있음.
    def __init__(self, name):
        self.name = name
    
    def update(self, message):
        print('{} got message "{}"'.format(self.name, message))

        
class Publisher:
    def __init__(self):
        self.observers = set()  # Publisher 객체 생성시 Subscribers를 담는 set 생성
    
    def register(self, who):
        self.observers.add(who)
        
    def unregister(self, who):
        self.observers.discard(who)
        
    def dispatch(self, message):   # subscriber의 update 함수 매개변수가 쓰임
        for observer in self.observers:
            observer.update(message)   # subscriber의 update 함수는 Publisher가 씀
            
if __name__ == "__main__":
    pub = Publisher()
    
    bob = Subscriber('Bob')
    alice = Subscriber('Alice')
    john = Subscriber('John')
    
    pub.register(bob)
    pub.register(alice)
    pub.register(john)
    
    # 공지하기(dispatch)
    pub.dispatch("It's lunchtime!")
    pub.unregister(john)
    pub.dispatch("Time for dinner")

Bob got message "It's lunchtime!"
Alice got message "It's lunchtime!"
John got message "It's lunchtime!"
Bob got message "Time for dinner"
Alice got message "Time for dinner"


In [7]:
# A Pythonic Refinement

# 기존의 Subscriber를 SubscriberOne과 SubscriberTwo로 나누어보자.
class SubscriberOne:
    def __init__(self, name):
        self.name = name
    
    def update(self, message):
        print('{} got message "{}"'.format(self.name, message))
        
        
class SubscriberTwo:
    def __init__(self, name):
        self.name = name
    
    def receive(self, message):  # SubscriberOne의 update와 기능 같음
        print('{} got message "{}"'.format(self.name, message))
        

class Publisher:
    def __init__(self):
        self.observers = dict()
    # register시 key:value=observer:callback 형식으로 저장
    # callback = observer.update(message) 정의 후 dispatch 함수에서는 callback(message)를 활용
    def register(self, who, callback=None): 
        if callback is None:
            callback = getattr(who, 'update')
        self.observers[who] = callback
        
    def unregister(self, who):
        del self.observers[who]
        
    def dispatch(self, message): # messasge
        for observer, callback in self.observers.items():
            callback(message)
            
if __name__ == "__main__":
    pub = Publisher()
    
    bob = SubscriberOne('Bob')
    alice = SubscriberTwo('Alice')
    john = SubscriberOne('John')
    
    pub.register(bob, bob.update)
    pub.register(alice, alice.receive) # alice는 update 메소드가 없어도 공지를 받을 수 있음(이 패턴 장점)
    pub.register(john)
    
    pub.dispatch("It's lunchtime!")
    pub.unregister(john)
    pub.dispatch("Time for dinner")

Bob got message "It's lunchtime!"
Alice got message "It's lunchtime!"
John got message "It's lunchtime!"
Bob got message "Time for dinner"
Alice got message "Time for dinner"


In [8]:
# Observing Events(notifying different group of subscribers for different situation)

class Subscriber:   # 모든 subscriber는 메세지를 받는 update 함수를 가지고 있음.
    def __init__(self, name):
        self.name = name
    
    def update(self, message):
        print('{} got message "{}"'.format(self.name, message))

        
class Publisher:
    def __init__(self, events):
        self.subscribers = { event : dict() for event in events}  # 이벤트별로 "이벤트: dict"를 만듬
        
    def get_subscribers(self, event):  # 이벤트별로 해당되는 subscribers 조회
        return self.subscribers[event]
    
    def register(self, event, who, callback=None):
        if callback is None:
            callback = getattr(who, 'update')
        self.get_subscribers(event)[who] = callback
        
    def unregister(self, event, who):
        del self.get_subscribers(event)[who]
        
    def dispatch(self, event, message): # messasge
        for subscriber, callback in self.get_subscribers(event).items():
            callback(message)
            
if __name__ == "__main__":
    pub = Publisher(['lunch', 'dinner'])
    
    bob = Subscriber('Bob')
    alice = Subscriber('Alice')
    john = Subscriber('John')
    
    pub.register('lunch', bob)
    pub.register('dinner', alice)
    pub.register('lunch', john)
    pub.register('dinner', john)
    
    pub.dispatch('lunch', "It's lunchtime!")
    pub.dispatch('dinner', "Dinner is served.")

Bob got message "It's lunchtime!"
John got message "It's lunchtime!"
Alice got message "Dinner is served."
John got message "Dinner is served."


# 6. Super() 기본 개념

## 6.1 super()

https://rednooby.tistory.com/56

- 자식 클래스에서 부모클래스의 내용을 사용하고 싶을경우 사용

In [9]:
class Father(): # 부모클래스
    def handsome(self):
        print("잘생겼다")
        
class Brother(Father): # 자식클래스 brother에서 부모클래스 father를 상속
    '''아들'''
    
class Sister(Father): # 자식클래스 sister에서 부모클래스 father를 상속
    def pretty(self):
        print("예쁘다")
        
    def handsome(self):
        '''물려 받았어요.'''
        

brother = Brother()
brother.handsome()

잘생겼다


In [11]:
girl = Sister()
girl.pretty()

예쁘다


In [12]:
girl.handsome() # 출력 안됨

In [13]:
class Father(): # 부모클래스
    def handsome(self):
        print("잘생겼다")
        
class Brother(Father): # 자식클래스 brother에서 부모클래스 father를 상속
    '''아들'''
    
class Sister(Father): # 자식클래스 sister에서 부모클래스 father를 상속
    def pretty(self):
        print("예쁘다")
        
    def handsome(self):
        super().handsome()

In [15]:
littleGirl = Sister()
littleGirl.handsome()

잘생겼다


In [16]:
# 응용

class Father():  # 부모 클래스
    def __init__(self, who):
        self.who = who
 
    def handsome(self):
        print("{}를 닮아 잘생겼다".format(self.who))
 
class Sister(Father):  # 자식클래스(부모클래스) 아빠매소드를 상속받겠다
    def __init__(self, who, where):
        super().__init__(who)
        self.where = where
 
    def choice(self):
        print("{} 말이야".format(self.where))
 
    def handsome(self):
       super().handsome()
       self.choice()

        
girl = Sister("아빠", "얼굴")
girl.handsome()

아빠를 닮아 잘생겼다
얼굴 말이야


## 6.2 super(Class, self)

https://hashcode.co.kr/questions/6419/python3-super%EC%99%80-supera-self%EC%9D%98-%EC%B0%A8%EC%9D%B4%EB%8A%94-%EB%AC%B4%EC%97%87%EC%9D%B8%EA%B0%80%EC%9A%94

- 위의 super()와 달리 괄호 안에 class명과 self가 매개변수로 들어가는 경우 존재
- 아래 예제처럼 super(B, self)라고 쓰면 B의 부모클래스를 상속받는다는 뜻임
- super(Class, self)는 상속받는 부모클래스가 여러개일 때 주로 사용

In [1]:
class A:
    def __init__(self):
        print("A")

class B(A):
    def __init__(self):
        super().__init__()

b = B()
b

A


In [2]:
class A:
    def __init__(self):
       print("A")

class B(A):
    def __init__(self):
        super(B, self).__init__()

b = B()
b

A


# 7. Queue 클래스

https://docs.python.org/3/library/queue.html

- multi-producer, multi-consumer queues를 개시하는데 쓰여 threaded programming에 활용이 자주됨
- 세가지 종류의 queue 존재
    1. FIFO(queue.Queue()): the first tasks added are the first retrieved.
    2. LIFO(queue.LifoQueue()): the most recently added entry is the first retrieved (operating like a stack)
    3. a priority queue(queue.PriorityQueue()): the entries are kept sorted and the lowest valued entry is retrieved first.

## Queue Objects

### Queue.qsize()
> return the approximate size of the queue

### Queue.empty()
> return **True** if the queue is empty, **False** otherwise.

### Queue.full()
> return **True** if the queue is full, **False** otherwise.

### Queue.put(_item, block=True, timeout=None_)
> - put item into the queue
> - if block=True, timeout=None, block if necessary until a free slot is available
> - if timeout=3, it block at most 3 seconds and raises the **Full** exception if no free slot was available within that time

### Queue.put_nowait(_item_)
> equivalent to put(item, False)

### Queue.get(_block=True, timeout=None_)
> - remove and return an item from the queue
> - If block=True, timeout=None, block if necessary until an item is available
> - If timeout is a positive number, it blocks at most timeout seconds and raises the Empty exception if no item was available within that time

### Queue.get_nowait()
> equivalent to get(False)

### Queue.task_done()
> For each get() used to fetch a task, a subsequent call to task_done() tells the queue that the processing on the task is complete.

### Queue.join()
> blocks until all items in the queue have been gotten and processed


In [3]:
import threading, queue

q = queue.Queue()

def worker():
    while True:
        item = q.get()
        print(f'Working on {item}')
        print(f'Finished {item}')
        q.task_done()

# turn-on the worker thread
threading.Thread(target=worker, daemon=True).start()

# send thirty task requests to the worker
for item in range(30):
    q.put(item)
print('All task requests sent\n', end='')

# block until all tasks are done
q.join()
print('All work completed')

All task requests sent
Working on 0
Finished 0
Working on 1
Finished 1
Working on 2
Finished 2
Working on 3
Finished 3
Working on 4
Finished 4
Working on 5
Finished 5
Working on 6
Finished 6
Working on 7
Finished 7
Working on 8
Finished 8
Working on 9
Finished 9
Working on 10
Finished 10
Working on 11
Finished 11
Working on 12
Finished 12
Working on 13
Finished 13
Working on 14
Finished 14
Working on 15
Finished 15
Working on 16
Finished 16
Working on 17
Finished 17
Working on 18
Finished 18
Working on 19
Finished 19
Working on 20
Finished 20
Working on 21
Finished 21
Working on 22
Finished 22
Working on 23
Finished 23
Working on 24
Finished 24
Working on 25
Finished 25
Working on 26
Finished 26
Working on 27
Finished 27
Working on 28
Finished 28
Working on 29
Finished 29
All work completed


# 8. thread에 대한 이해

링크: https://www.youtube.com/watch?v=IEEhzQoKtQU

먼저, threading 없이 do_something() 이라는 함수를 두번 시행하는데 걸리는 시간을 측정해보자.

In [1]:
import time

start = time.perf_counter()

def do_something():
    print('Sleeping 1 second...')
    time.sleep(1)
    print('Done Sleeping...')

do_something()
do_something()

finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')

# Finished in 2.0 second(s)

Sleeping 1 second...
Done Sleeping...
Sleeping 1 second...
Done Sleeping...
Finished in 2.0 second(s)


do_something()에 1초간 쉬라는 명명의 time.sleep(1)이 있어 do_something()을 수행하는데 1초 가량이 걸리고, do_something()이 두번 수행되므로 총 2초 가량이 걸림을 알 수 있다.

위 코드에 threading을 적용 해 보자.

In [2]:
import threading
import time

start = time.perf_counter()

def do_something():
    print('Sleeping 1 second...')
    time.sleep(1)
    print('Done Sleeping...')

t1 = threading.Thread(target=do_something)
t2 = threading.Thread(target=do_something)

t1.start()
t2.start()

finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')

# Finished in 0.0 second(s) 라고 뜬 후 Done Sleeping... 이 프린트됨
# reason : while the tread is for sleeping, our script ran concurrently and continued on with the rest of the script

Sleeping 1 second...
Sleeping 1 second...Finished in 0.02 second(s)

Done Sleeping...Done Sleeping...



- 위 코드를 보면 threading.Thread(target="함수명")을 통해 t1이라는 스레드를 구성 후, t1.start() 명령어를 통해 함수를 실행
- 하지만 함수 실행이 완료(Done Sleeping...)되기도 전에 코드의 가장 마지막 명령어(Finished in 0.02 second(s))가 수행됨
- 이는 스레드가 쉬는 동안 아래 부분 코드를 동시에 읽어들여 수행하기 때문
- 스레드를 끝내기 위해선 t1.join()이라는 명령어가 필요(아래 코드 참조)

In [3]:
import threading
import time

start = time.perf_counter()

def do_something():
    print('Sleeping 1 second...')
    time.sleep(1)
    print('Done Sleeping...')

t1 = threading.Thread(target=do_something)
t2 = threading.Thread(target=do_something)

t1.start()
t2.start()

t1.join() # complete the running thread
t2.join()

finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')

Sleeping 1 second...
Sleeping 1 second...
Done Sleeping...
Done Sleeping...
Finished in 1.01 second(s)


- finish = time.perf_counter() 명령어 전 t1.join() 명령어를 넣어 작동 중인 스레드를 종료 시킴
- 2개의 스레드를 만들고 각 스레드에 do_something() 함수를 동시에 처리한 결과 2초가 걸리는 작업을 1.01초로 단축 시켰음


- 아래는 for문을 이용하여 10개의 스레드를 만들어 do_something()을 병렬처리하는 코드임.
- 10번의 do_something() 처리엔 10초가 걸려야 하지만 스레드를 통한 병렬처리로 1.03초만에 해결함

In [6]:
import threading
import time

start = time.perf_counter()

def do_something():
    print('Sleeping 1 second...')
    time.sleep(1)
    print('Done Sleeping...')

threads = []

for _ in range(10):
    t = threading.Thread(target=do_something)
    t.start()
    threads.append(t)

for thread in threads:
    thread.join()

finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')

# Finished in 1.03 second(s)

Sleeping 1 second...
Sleeping 1 second...
Sleeping 1 second...
Sleeping 1 second...
Sleeping 1 second...Sleeping 1 second...
Sleeping 1 second...

Sleeping 1 second...
Sleeping 1 second...
Sleeping 1 second...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Finished in 1.03 second(s)


- do_something 함수에 parameter가 존재한다면, t = threading.Thread(target="함수명", args=list(변수명))을 써주면 된다

In [7]:
import threading
import time

start = time.perf_counter()

def do_something(seconds):
    print(f'Sleeping {seconds} second(s)...')
    time.sleep(seconds)
    print('Done Sleeping...')

threads = []

for _ in range(10):
    t = threading.Thread(target=do_something, args=[1.5])
    t.start()
    threads.append(t)

for thread in threads:
    thread.join()

finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')

# Finished in 1.54 second(s)

Sleeping 1.5 second(s)...
Sleeping 1.5 second(s)...
Sleeping 1.5 second(s)...
Sleeping 1.5 second(s)...
Sleeping 1.5 second(s)...
Sleeping 1.5 second(s)...
Sleeping 1.5 second(s)...
Sleeping 1.5 second(s)...
Sleeping 1.5 second(s)...
Sleeping 1.5 second(s)...
Done Sleeping...
Done Sleeping...
Done Sleeping...Done Sleeping...

Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Finished in 1.54 second(s)


- 이번에는 threading 모듈 대신 concurrent.futures 모듈을 사용해보자
- concurrent.futures를 사용하기 위해선 target 함수에 반드시 return 값이 존재해야 함
- 스레드 개시는 executor.submit("함수명", "변수"), 스레드 종료는 f1.result()

In [8]:
import concurrent.futures
import time

start = time.perf_counter()

def do_something(seconds):
    print(f'Sleeping {seconds} second(s)...')
    time.sleep(seconds)
    return 'Done Sleeping...'

with concurrent.futures.ThreadPoolExecutor() as executor:
    f1 = executor.submit(do_something, 1) # f1 is futures object. 
    f2 = executor.submit(do_something, 1)

    print(f1.result())
    print(f2.result())

finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')

# Finished in 1.02 second(s)

Sleeping 1 second(s)...Sleeping 1 second(s)...

Done Sleeping...
Done Sleeping...
Finished in 1.02 second(s)


- 이번에는 concurrent.futures에 for문을 적용하여 10개의 함수를 동시에 돌려보자

In [9]:
import concurrent.futures
import time

start = time.perf_counter()

def do_something(seconds):
    print(f'Sleeping {seconds} second(s)...')
    time.sleep(seconds)
    return 'Done Sleeping...'

with concurrent.futures.ThreadPoolExecutor() as executor:
    results = [executor.submit(do_something, 1) for _ in range(10)]
    
    for f in concurrent.futures.as_completed(results):
        print(f.result())

finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')

# Finished in 1.05 second(s)

Sleeping 1 second(s)...Sleeping 1 second(s)...

Sleeping 1 second(s)...
Sleeping 1 second(s)...
Sleeping 1 second(s)...
Sleeping 1 second(s)...
Sleeping 1 second(s)...
Sleeping 1 second(s)...
Sleeping 1 second(s)...
Sleeping 1 second(s)...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Done Sleeping...
Finished in 1.05 second(s)


- 아래처럼 각 스레드별로 변수를 다르게 넣을 수도 있다

In [11]:
import concurrent.futures
import time

start = time.perf_counter()

def do_something(seconds):
    print(f'Sleeping {seconds} second(s)...')
    time.sleep(seconds)
    return f'Done Sleeping...{seconds}'

with concurrent.futures.ThreadPoolExecutor() as executor:
    secs = [5, 4, 3, 2, 1]
    results = [executor.submit(do_something, sec) for sec in secs]
    
    for f in concurrent.futures.as_completed(results):
        print(f.result())

finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')

# Finished in 5.01 second(s)

Sleeping 5 second(s)...
Sleeping 4 second(s)...
Sleeping 3 second(s)...
Sleeping 2 second(s)...
Sleeping 1 second(s)...
Done Sleeping...1
Done Sleeping...2
Done Sleeping...3
Done Sleeping...4
Done Sleeping...5
Finished in 5.01 second(s)


In [9]:
import concurrent.futures
import time

start = time.perf_counter()

def do_something(seconds):
    print(f'Sleeping {seconds} second(s)...')
    time.sleep(seconds)
    return f'Done Sleeping...{seconds}'

with concurrent.futures.ThreadPoolExecutor() as executor:
    secs = [5, 4, 3, 2, 1]
    results = executor.map(do_something, secs)

    for result in results:
        print(result)


finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')

# Finished in 5.01 second(s)

Sleeping 5 second(s)...
Sleeping 4 second(s)...
Sleeping 3 second(s)...
Sleeping 2 second(s)...
Sleeping 1 second(s)...
Done Sleeping...5
Done Sleeping...4
Done Sleeping...3
Done Sleeping...2
Done Sleeping...1
Finished in 5.01 second(s)


- 이번에는 threading.Condition()의 wait()과 notifyAll() 함수를 이용하여 스레드를 제어하는 방법을 알아보자
- threading.Condition.wait()은 돌고있는 스레드를 잠시 멈추게 하고, threading.Condition.notifyAll()은 잠시 멈춘 스레드를 다시 동작하게 한다.

In [12]:
import threading

def consumer(cond):
    name = threading.currentThread().getName()
    print("{0} 시작".format(name))
    with cond:
        print("{0} 대기".format(name))
        cond.wait()
        print("{0} 자원 소비".format(name))

def producer(cond):
    name = threading.currentThread().getName()
    print("{0} 시작".format(name))
    with cond:
        print("{0} 자원 생산 후 모든 소비자에게 알림".format(name))
        cond.notifyAll()


if __name__ == "__main__":
    condition = threading.Condition()
    consumer1 = threading.Thread(
        name="소비자1", target=consumer, args=(condition,)
    )
    consumer2 = threading.Thread(
        name="소비자2", target=consumer, args=(condition,)
    )
    producer = threading.Thread(
        name="생산자", target=producer, args=(condition,)
    )

    consumer1.start()
    consumer2.start()
    producer.start()


소비자1 시작
소비자1 대기
소비자2 시작
소비자2 대기
생산자 시작
생산자 자원 생산 후 모든 소비자에게 알림
소비자1 자원 소비
소비자2 자원 소비


# 9. 다시 봐도 헷갈리는 비트연산자

링크: https://wikidocs.net/1161

- 비트연산자는 십진수를 이진수 기준으로 연산하여 값을 돌려주는 것이 기본
- a = 60, b = 13 이라고 한다면
- a(2) = 0011 1100
- b(2) = 0000 1101

In [1]:
a, b = 60, 13

In [2]:
# & : and 연산(교집합), 둘다 1일 때만 1로 표시
a & b
# 12(2) = 0000 1100

12

In [3]:
# | : or 연산(합집합), 둘 중 하나만 1이어도 1로 표시
a | b
# 61(2) = 0011 1100

61

In [4]:
# ^ : xor 연산(합집합 - 교집합), 둘 중 하나만 1일 때 1로 표시
a ^ b
# 49(2) = 0011 0001

49

## 보수 연산(~)

링크1: https://abn-abn.tistory.com/17 <br>
링크2: https://abn-abn.tistory.com/17 <br>
링크3: https://blog.kimtae.xyz/5 <br>

### 보수란?
- 주어진 수와 더했을 때 0이 되는 수
- 즉, 10의 보수는 -10, 2의 보수는 -2. 따라서 10진수에서 보수는 부호만 바뀐값

### 2진수에서 어떤 수(A)의 보수(-A) 표현
- 양의 정수 A의 보수인 음의 정수 -A로 바꾸기 위해서는 0은 1로, 1은 0으로 바꾼 뒤, 1을 더하면 된다
- 예, 12(2) = 0000 1100의 보수 -12(2)는 1111 0011에서 1을 더한 1111 0100이 됨
- 위와 같이 해야 주어진 수와 보수의 합 계산시 0이 되기 때문. 즉, 12(2) + (-12(2)) = 0
-  12(2) =   0000 1100
- -12(2) =   1111 0100
-   0(2) = 1 0000 0000
- 최종 결과에서 부호를 결정하는 앞자리수만 1이 되고 나머지는 0이 된다. 1 Byte는 8Bit이므로 앞자리 1은 자리가 없어 삭제됨

In [5]:
# ~ : 보수 연산, a라는 값 주어질 때 -a값 구함
~a
# a(2)       = 0011 1100
# a(2) 뒤집기 = 1100 0011
# 뒤집은 a(2)에 +1 = 1 0000 0000

-61

In [6]:
~(~a)

60

In [11]:
a + (~a)

-1

In [12]:
# a << 2 : 왼쪽 시프트 연산자. 변수값을 왼쪽으로 지정된 비트 수만큼 이동
# 0011 1100 -> 1111 0000 = 240
a << 2

240

In [13]:
# a >> 2 : 오른쪽 시프트 연산자. 변수값을 오른쪽으로 지정된 비트 수만큼 이동
# 0011 1100 -> 0000 1111 = 15
a >> 2

15

# 10. 논리 연산자(and, or, not)

링크1: https://dojang.io/mod/page/view.php?id=2192

## a and b
- True/False : and는 두값이 모두 True이어야 True (False preferred)
- Bool과 숫자 : False and 숫자 -> False, True and 숫자 -> 숫자
- 숫자끼리 : 3 and 4 -> 뒤에 있는 숫자인 4를 프린트(대소 상관 없는 듯)

In [1]:
True and True

True

In [2]:
True and False

False

In [3]:
False and True

False

In [4]:
False and False

False

In [5]:
False and 2

False

In [8]:
True and 2

2

In [7]:
3 and 4

4

In [9]:
45 and 22

22

In [10]:
45 and 33

33

## a or b
- True/False : or은 두값 중 하나라도 True이면 True (True preferred)
- Bool과 숫자 : False or 숫자 -> 숫자, True or 숫자 -> 숫자
- 숫자끼리 : 3 or 4 -> 앞에 있는 숫자인 4를 프린트(대소 상관 없는 듯)

In [11]:
True or True

True

In [12]:
True or False

True

In [13]:
False or True

True

In [14]:
False or False

False

In [15]:
True or 23

True

In [16]:
23 or True

23

In [17]:
False or -1

-1

In [18]:
2 or False

2

In [19]:
3 or 4

3

In [20]:
4 or 2

4

In [23]:
22 or 42

22

## not a
- Bool : not False -> True, not True -> False
- not 숫자 : False

In [24]:
not False

True

In [25]:
not True

False

In [26]:
not 3

False

In [27]:
not 22

False